### PR TITLE
Add hover state to Tree items display

### PR DESCRIPTION
**Pull request by davthedev**
Implements https://github.com/godotengine/godot-proposals/issues/7160https://github.com/godotengine/godot-proposals/issues/7160

Since this is a bit less trivial than other hover additions, here are some explanations.

This change aims at improving the usability of the scene tree and all places where a Tree control is used. Moving the mouse above a tree item highlights it. 

The highlight is also rendered on individual hovered icon buttons. In this case, the main row or cell it belongs to keeps its highlight in a "dimmed" version. This helps focusing on the icon while keeping the visual relationship with the piece of data the action is likely to impact.

Example with a slightly customized theme to illustrate a dense list:

itemshover

Detail on the "dimmed hover" effect:

itemshover1

To make it more obvious, here it is with customized colors:

itemshover2

Projects can take advantage of the new styling possibilities. All new visual elements are customizable using appropriate theme resources.

Editor and default themes are adjusted to provide the additional theme resources. Other existing stuff is left untouched in this regard, so that there are no regressions on existing projects.

The Tree control is a complex widget, handling many possibilities of display. The display and original click handling algorithm is based on recursive traversal.

I implemented a direct, lighter way to find out which item is hovered that exactly matches, or at best attempt, the algorithm used to determine the clicked item, because mouse movemens happen more often.

This PR is exclusively adding support for hover feedback. It does not fix a few inconsistencies I found in click coordinates resolution, especially for the buttons attached to a cell (think the script, show/hide, etc... you see in the scene tree) with some combination of spacings. Those may be addressed in separate PRs. The goal here is that the hovered zones match the currently clickable zones.

A good part of the complexity is that there is an algorithm to move those icons towards the left in comparison to their original position in the cell if there is horizontal scrolling. This is what keeps the icon buttons visible in the scene tree at all times even if there is a long node name that overflows. The logic to compute that offset is a little complex and I had to replicate it. The said algorithm is not implemented in right-to-left display and some of the public methods that return the position of the in-cell buttons do not apply it - matter for separate issue reports.

### CODE REVIEW

File: doc/classes/Tree.xml
Comment (Mickeon) [inline doc/classes/Tree.xml:517]: Unsure of what this is referring to, when looking at the video preview and description. Calling the property itself "dimmed" is not a good idea. The adjective refers to the custom color used, and as such it may not always be _dimmed_.

File: doc/classes/Tree.xml
Comment (Mickeon) [inline doc/classes/Tree.xml:683]: Ditto.

File: doc/classes/Tree.xml
Comment (davthedev) [reply doc/classes/Tree.xml:517]: When you hover the action icons, the row still has a hover highlight but different than when you hover the "main" part of the row, as I explained above. 

Observe the animation closely and you will spot the difference, when the mouse moves between the main part of the row and the action icons **on an unselected row**. The row highlight is still here, with less contrast in both the selection box and the text color (hence the two properties you mention). Of course, the user is free to set anything for those and they can be rendered using a completely different color; not just a dimmed aspect.

This allows the icons hover to be visible even if the Tree is in full row highlight mode.

I had to choose a word. Would you have a different suggestion? Ideas: hovered_secondary

Comment (Mickeon) [reply doc/classes/Tree.xml:517]: The current color should've been called `font_hovered_highlight_color` and this one should be `font_hovered_color` but that obviously cannot be done, unless we're doing some very messy compatibility work, hmm.... That's a pretty difficult question, actually.

- Anything with "focus" is out of the question as the concept applies to something else outright for Controls.
- Adjectives like "faint", "low" are not that good for the same reasons as "dimmed"

If I have a saying on this, I would call it based on the very thing it is used for, no vague adjectives: `font_hovered_on_button_color`. Something like this.

Comment (davthedev) [reply doc/classes/Tree.xml:517]: "Hovered on button" may lead to confusion, because there is no mechanism to customize the buttons hover color yet. 

The naming question should be taken globally for the font (font_hovered_dimmed_color) and the background (hovered_dimmed) at the same time. If we attack problems one by one without seeing the bigger picture, we lose consistency.

**Review by Calinou — COMMENTED** [review]
Tested locally, it works as expected. This makes Tree-based controls a lot nicer to use, great work :slightly_smiling_face: 

### Black (OLED) theme

https://github.com/godotengine/godot/assets/180032/53af857c-6727-4a23-9f38-72108e0b311d

### Light theme

https://github.com/godotengine/godot/assets/180032/abc638df-000b-4912-a7a7-ed32f256f080

I've ensured that continuous redrawing doesn't occur if the mouse stands still on a TreeItem by looking at the update spinner.

Code looks good to me at a glance, aside of the style suggestions.

File: scene/gui/tree.cpp
Comment (Calinou) [inline scene/gui/tree.cpp:4931]: We use snake_case only in Godot:

```suggestion
	int nesting_level = 0;
```

File: scene/gui/tree.cpp
Comment (Calinou) [inline scene/gui/tree.cpp:4939]:
```suggestion
				nesting_level += 1;
```

File: scene/gui/tree.cpp
Comment (Calinou) [inline scene/gui/tree.cpp:4942]:
```suggestion
				nesting_level -= 1;
```

File: scene/gui/tree.cpp
Comment (Calinou) [inline scene/gui/tree.cpp:4949]:
```suggestion
			nesting_level += 1;
```

File: scene/gui/tree.cpp
Comment (Calinou) [inline scene/gui/tree.cpp:4956]:
```suggestion
				nesting_level -= 1;
```

File: scene/gui/tree.cpp
Comment (Calinou) [inline scene/gui/tree.cpp:4944]:
```suggestion
			return nesting_level * theme_cache.item_margin;
```

File: doc/classes/Tree.xml
Comment (Calinou) [reply doc/classes/Tree.xml:517]: I think `dimmed` is fine here (`muted` would work as a synonym, which is used in Material UI guidelines IIRC). It's hard to convey with a single word what it's exactly supposed to do.

File: scene/gui/tree.cpp
Comment (KoBeWi) [inline scene/gui/tree.cpp:2917]:
```suggestion
		// Interesting cell button detection code.
```

File: scene/gui/tree.cpp
Comment (KoBeWi) [inline scene/gui/tree.cpp:4988]:
```suggestion
	return -1; // Not found.
```

**Review by KoBeWi — APPROVED** [review]
Code and the functionality itself seems fine, though I wonder if some parts can't be simplified (like the button hover code; since buttons are clickable, it should be detected elsewhere, no?)

**Comment by davthedev** [comment]
This complex control has to be refactored in order to fix some positioning quirks. This waits for a more important upgrade where it is admissible to potentially break some themes.

For the changes covered in this PR, I reproduced in the hover code the way side buttons are positioned including the quirks. For example, you can see those giggling sideways if you scroll the inspector horizontally when names are longer. The original click detection is a recursive tree traversal; for the hover I found a simpler way without having to hook up to this recursivity. That is less execution complexity as well for an event that occurs more often than clicks. As some refactoring is planned for later, I opted of the simpler solution.

**Comment by akien-mga** [comment]
CC @passivestar as I believe you had Opinions™ on this type of hover/select UX ;)

**Comment by passivestar** [comment]
> CC @passivestar as I believe you had Opinions™ on this type of hover/select UX ;)

I like it, I think godot needs it. I actually wanted to work on this myself before I discovered this PR

I think I found a bug where it sometimes won't unhover when the mouse leaves the dock:

https://github.com/godotengine/godot/assets/60579014/f9e38469-ac2b-4cc4-8ee3-ce8fa449dee0

File: scene/gui/tree.cpp
Comment (AThousandShips) [inline scene/gui/tree.cpp:3687]:
```suggestion
				// Find possible hovered button in cell.
```
Eventual = sooner or later

**Comment by akien-mga** [comment]
> I think I found a bug where it sometimes won't unhover when the mouse leaves the dock:

@davthedev Would you have time to see if you can reproduce and fix that issue?

**Comment by davthedev** [comment]
Sure! I will have more time to have a look in a couple days.

**Comment by passivestar** [comment]
Tree seems to be getting an `InputEventMouseMotion` after `NOTIFICATION_MOUSE_EXIT`, causing `cache.hover_item` to be set again before the queued redraw, which causes the bug that I found

You could either delay nullifying `cache.hover_item` to make sure it's null before the redraw happens, or properly check that the mouse cursor is in the tree when we're setting things in `gui_input`

None of that seems ideal though and I find it weird that we're getting a mouse motion event after the mouse exit notification, is it the intended behavior in godot?

**Comment by davthedev** [comment]
Also noted another couple bugs I should fix:

- If the content is refreshed while hovering, the highlight is lost until the next mouse move.
This happens on the "Distant" nodes tree while a scene is running.

- Hover highlight is also visible during a drag-drop operation and can be misleading at that moment, as the one thing that should be highlighted during this operation is the drop target.

**Comment by davthedev** [comment]
Just fixed the two bugs above. Changed the mouse over (or not) tracking logic to tentatively fix @passivestar hover feedback bug, relying on MOUSE_ENTER / _EXIT notifications. As I was not able to reproduce on my side, can you have a look?

**Comment by passivestar** [comment]
> As I was not able to reproduce on my side, can you have a look?

The bug is gone 🥳

I noticed two other things though:

1. Hovering in the project manager is broken now? It uses a bright thin outline instead of whatever is was before

https://github.com/godotengine/godot/assets/60579014/0d0f5b35-f9ce-4727-8b56-21e6425061b1

2. When hovering in a tree that has more than one main non-button column it highlights the right column, which doesn't always make sense semantically. In the shortcuts editor for example those buttons are related to the entire row, not just the rightmost column. When using trees in games it might not always make sense either

https://github.com/godotengine/godot/assets/60579014/e2578f49-6162-42ea-9ab5-5254a155049d

Perhaps when a tree has more than 2 main columns and you're hovering over a button it should highlight ALL of the columns or NONE of them, instead of the right one?

**Comment by davthedev** [comment]
> I noticed two other things though:
> 
>     1. Hovering in the project manager is broken now? It uses a bright thin outline instead of whatever is was before

Looks like a broken editor theme. The thin grey outline is the default StyleBox when nothing is specified in the system theme; I will have a look. I rebased my month-old work with lots of commits in between, probable collateral damage...

> 
>     2. When hovering in a tree that has more than one main non-button column it highlights the right column, which doesn't always make sense semantically. In the shortcuts editor for example those buttons are related to the entire row, not just the rightmost column. When using trees in games it might not always make sense either
> 

Because there are multiple possible select modes in the Tree control that can be set on each instance: entire row or single cell. The view you are referring to is in single-cell mode, unlike the node tree. It works like a spreadsheet editor. I think it is up to this specific screen to work its way around instead by configuring the Tree control instance correctly.

Here, the cell is the actual binding, which you can edit using the dedicated buttons. Makes sense IMHO.

If some actions need to relate to the whole row, while keeping the separate cell selection possible, there are multiple options:

- Configure this screen in single-row selection mode - check UX as, for now, double-clicking the left cell unfolds the tree, and double-clicking the right cell opens the binding edit dialog.
- Make a separate third column for the buttons
- Add a setting to the Tree control mandating to highlight the whole row on specific cells / columns or buttons. Or a "select single cell but highlight whole row on buttons" selection mode

**Comment by Calinou** [comment]
@davthedev Could you look into rebasing this PR on top of `master`? I'd like to give this a final spin so we can look into merging this for 4.4.

**Comment by davthedev** [comment]
Rebasing done, but wait for my green light before merging. I must double check that previously mentioned side effects are resolved. The one with the project chooser window is kind of critical.

**Comment by Calinou** [comment]
> Rebasing done, but wait for my green light before merging. I must double check that previously mentioned side effects are resolved. The one with the project chooser window is kind of critical.

The issue is still present, I just tested:

image

The rest seems to work fine.

**Comment by davthedev** [comment]
Theme issue in the project selector should be fixed now. Good to merge!

**Review by Calinou — APPROVED** [review]
Tested again, it works great now, including in the project manager :slightly_smiling_face:

File: scene/gui/tree.cpp
Comment (AThousandShips) [inline scene/gui/tree.cpp:4189]:
```suggestion
					// Only need to update if mouse enters/exits a button.
```

File: scene/gui/tree.cpp
Comment (AThousandShips) [inline scene/gui/tree.cpp:3012]:
```suggestion
		// Interesting cell button detection code.
```
Also interesting how? What is this comment saying?

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**
